Userland/Libraries/LibGUI/ZoomableView.cpp

```diff
@@ -0,0 +1,179 @@
+    Gfx::IntSize new_size;
+    new_size.set_width(m_original_rect.width() * m_scale);
+    new_size.set_height(m_original_rect.height() * m_scale);
+    m_image_rect.set_size(new_size);
+
+    if (on_scale_change)
+        on_scale_change(m_scale);
+
+    relayout();
+}
+
+void ZoomableView::scale_by(float delta)
+{
+    float new_scale = m_scale * AK::exp2(delta);
+    set_scale(new_scale);
+}
+
+void ZoomableView::scale_centered(float new_scale, Gfx::IntPoint const& center)
+{
+    if (m_original_rect.is_null())
+        return;
+
+    new_scale = max(min(new_scale, m_max_scale), m_min_scale);
+
+    if (new_scale == m_scale)
+        return;
+
+    Gfx::FloatPoint focus_point {
+        center.x() - width() / 2.0f,
+        center.y() - height() / 2.0f
+    };
+    m_origin = (m_origin + focus_point) * (new_scale / m_scale) - focus_point;
+    set_scale(new_scale);
+}
+
+void ZoomableView::start_panning(Gfx::IntPoint const& position)
+{
+    m_saved_cursor = override_cursor();
+    set_override_cursor(Gfx::StandardCursor::Drag);
+    m_pan_start = m_origin;
+    m_pan_mouse_pos = position;
+    m_is_panning = true;
+}
+
+void ZoomableView::stop_panning()
+{
+    m_is_panning = false;
+    set_override_cursor(m_saved_cursor);
+}
+
+void ZoomableView::pan_to(Gfx::IntPoint const& position)
+{
+    auto delta = position - m_pan_mouse_pos;
+    m_origin = m_pan_start.translated(-delta.x(), -delta.y());
+    relayout();
+}
+
+Gfx::FloatPoint ZoomableView::frame_to_image_position(Gfx::IntPoint const& frame_position) const
+{
+    Gfx::FloatPoint image_position;
+    image_position.set_x(((float)frame_position.x() - (float)m_image_rect.x()) / m_scale);
+    image_position.set_y(((float)frame_position.y() - (float)m_image_rect.y()) / m_scale);
+    return image_position;
+}
+
+Gfx::FloatRect ZoomableView::frame_to_image_rect(Gfx::IntRect const& frame_rect) const
+{
+    Gfx::FloatRect image_rect;
+    image_rect.set_location(frame_to_image_position(frame_rect.location()));
+    image_rect.set_width((float)frame_rect.width() / m_scale);
+    image_rect.set_height((float)frame_rect.height() / m_scale);
+    return image_rect;
+}
+
+Gfx::FloatPoint ZoomableView::image_to_frame_position(Gfx::IntPoint const& image_position) const
+{
+    Gfx::FloatPoint frame_position;
+    frame_position.set_x(m_image_rect.x() + ((float)image_position.x() * m_scale));
+    frame_position.set_y(m_image_rect.y() + ((float)image_position.y() * m_scale));
+    return frame_position;
+}
+
+Gfx::FloatRect ZoomableView::image_to_frame_rect(Gfx::IntRect const& image_rect) const
+{
+    Gfx::FloatRect frame_rect;
+    frame_rect.set_location(image_to_frame_position(image_rect.location()));
+    frame_rect.set_width((float)image_rect.width() * m_scale);
+    frame_rect.set_height((float)image_rect.height() * m_scale);
+    return frame_rect;
+}
+
+void ZoomableView::mousewheel_event(GUI::MouseEvent& event)
+{
+    float new_scale = scale() / AK::exp2(event.wheel_delta() / 8.0f);
```

Comment:
The magic number `8.0f` could go into a `constexpr int wheel_zoom_facor = 8.0f` at the top of the file as to not hide in this function.
I think this is the more serenity way...